skill-tree /src/main/java/com/RDS/skilltree/Skill/SkillsServiceImpl.java

```diff
@@ -0,0 +1,64 @@
+package com.RDS.skilltree.Skill;
+
+import com.RDS.skilltree.User.UserModel;
+import com.RDS.skilltree.User.UserRepository;
+import lombok.NoArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.dao.DataIntegrityViolationException;
+import org.springframework.data.domain.Page;
+import org.springframework.data.domain.Pageable;
+import org.springframework.stereotype.Service;
+
+import java.sql.SQLIntegrityConstraintViolationException;
+import java.time.Instant;
+import java.util.List;
+import java.util.Optional;
+import java.util.UUID;
+import java.util.stream.Collectors;
+
+@Service
+@Slf4j
+public class SkillsServiceImpl implements SkillsService{
+    private final SkillRepository skillRepository;
+    private final UserRepository userRepository;
+    public SkillsServiceImpl(SkillRepository skillRepository,
+                             UserRepository userRepository){
+        this.userRepository = userRepository;
+        this.skillRepository = skillRepository;
+    }
+
+    @Override
+    public SkillDTO getSkillById(UUID id){
+        Optional<SkillModel> skillModel = skillRepository.findById(id);
+        return skillModel.map(SkillDTO::toDto).orElse(null);
+    }
+
+    @Override
+    public SkillDTO  getSkillByName(String skillName){
+        Optional<SkillModel> skillModel = skillRepository.findByName(skillName);
+        return skillModel.map(SkillDTO::toDto).orElse(null);
+    }
+
+    @Override
+    public Page<SkillDTO> getAllSkills(Pageable pageable){
+        Page<SkillModel> skillModels = skillRepository.findAll(pageable);
+        return skillModels.map(SkillDTO::toDto);
+    }
+
+    @Override
```

Comment:
here it appears to be handling skill creation using a SkillDTO and SkillModel objects. However, there's a potential bug in the catch block that might cause issues with handling data integrity violations.

The issue lies in the fact that the exception is being rethrown without any additional handling. This means that if any data integrity violation occurs during the saving process, the exception will bubble up and potentially cause the application to crash.

To improve error handling and prevent potential crashes, consider implementing a more graceful approach:

Instead of directly re-throwing the exception, log the error details and return an appropriate response to the caller, indicating that skill creation failed due to a data integrity issue.

Consider implementing a custom exception class specifically for handling skill creation errors. This would allow for more specific error handling and better communication of the issue to the caller.

If the data integrity violation is related to a unique constraint, such as a duplicate skill name, consider returning a specific error message indicating that the skill name already exists. This would provide clearer guidance to the user on how to resolve the issue.

By implementing more robust error handling, the application can handle data integrity violations gracefully and provide meaningful feedback to the user, preventing unexpected crashes and improving the overall user experience.

Reply:
We are catching this exception in specific in the controller layer and returning the appropriate message [line ref](https://github.com/Real-Dev-Squad/skill-tree-backend/pull/34/files#diff-6b4e9eeeec93350cc7aa4c0456b785963790da16f2665cc5096352ce75ce93ffR36), we just caught it to have the exception logged in case of debugging.

Reply:
We can consider using aspect oriented programming by implementing libraries like AspectJ for logging.

Reply:
I used Slf4j for logging the error messages. @Ajeyakrishna-k